### PR TITLE
Adds climate driven planting date for crops

### LINKED ISSUE #2962: Dynamic Landunits won't work when a new crop is introduced
Crop land units require a 1-year spinup to calculate the growing season length. With dynamic landunits turned on, new crops introduced to a grid cell that were not present previously will not begin this counter and therefore will not be planted.

### PR DESCRIPTION
**Pull request by bbye**
Introduces a new method to calculate crop planting dates. Calculates the
temperature or precipitation seasonality of a grid and determines planting
date from temperature or precipitation thresholds. Therefore, the growing
season is either the warm or wet season. This allows crops to grow globally,
rather than in just the temperate latitudes.

A new module, CropMod is added to support some of the calculations and
will serve as a home of future crop model developments that are specific
to crops. The parameter file is not changed as there are no new parameters.

Also, fixes the dynamic landunit bug that won't allow crops to be planted
if they weren't on the grid cell previously.

Fixes #2962
[Non-BFB] - Non Bit-For-Bit (only for the crop model)

### CODE REVIEW

**Comment by bbye** [comment]
The design document can be found at: https://acme-climate.atlassian.net/wiki/spaces/ED/pages/994934933/B5+Crop+Plant+Date+Design+Document

**Comment by bbye** [comment]
The baseline comparison tests for e3sm_land_developer all pass on anvil, with the exception of the error TPUTCOMP Error: Computation time increase > 10 pct from baseline, which I was told to ignore.

**Comment by bishtgautam** [comment]
@bbye Is the new physics covered by the existing crop test? Or, do we need to add an additional test for the new physics?

**Comment by bbye** [comment]
The current test will be fine. Once the surface datasets are updated I can create an additional test that would be a more thorough check, but that's not necessary.

File: components/clm/src/biogeochem/CNPhenologyBeTRMod.F90 [hunk 1]
```diff
@@ -1347,6 +1356,14 @@ subroutine CropPhenology(num_pcropp, filter_pcropp                     , &
     real(r8) dayspyr  ! days per year
     real(r8) crmcorn  ! comparitive relative maturity for corn
     real(r8) ndays_on ! number of days to fertilize
+    logical p_season  ! precipitation seasonal
+    logical t_season  ! temperature seasonal
+    logical no_season ! neither temperature or precipitation seasonal
+    real(r8), parameter :: tcvp = 0.4       ! threshold of CV of precipitation to indicate seasonality
```
Comment (bishtgautam) [inline components/clm/src/biogeochem/CNPhenologyBeTRMod.F90]: Seems like you are defining `tcvp` and `tcvt` also in https://github.com/E3SM-Project/E3SM/pull/2999/files#diff-ac10333178760bf1fc41e8763b73a080R114. Is it possible to consolidate all those definition in a single module and include them via `use <module>`?

File: components/clm/src/biogeochem/CNPhenologyBeTRMod.F90 [hunk 2]
```diff
@@ -1542,45 +1576,78 @@ subroutine CropPhenology(num_pcropp, filter_pcropp                     , &
...
-                  if (ivt(p)==ncorn .or. ivt(p)==ncornirrig) then
-                     gddmaturity(p)=max(950._r8, min(gdd820(p)*0.85_r8, hybgdd(ivt(p))))
-                     gddmaturity(p)=max(950._r8, min(gddmaturity(p)+150._r8,1850._r8))
+               ! Assign seasonality - either temperature or precipitation
+               t_season = .false.
+               p_season = .false.
+               no_season = .false.
+               if (cvp(p) > tcvp) then        ! precipitation CV indicator
+                  if (cvt(p) >= tcvt) then    ! Both temperature and precip seasonality
+                     if (minval(xt_bar(p,:)) .lt. 283._r8)then ! cold season exists (i.e., min temp below 10 deg C)
```
Comment (bishtgautam) [inline components/clm/src/biogeochem/CNPhenologyBeTRMod.F90]: consider defining a variable for `283._r8`.

File: components/clm/src/biogeochem/PhenologyMod.F90 [hunk 3]
```diff
@@ -1592,45 +1626,78 @@ subroutine CropPhenology(num_pcropp, filter_pcropp                     , &
...
+                     idop(p)      = jday
+                     plantday(p)  = jday
+                     harvday(p)   = NOT_Harvested
+                     harvdate(p)  = NOT_Harvested
+
+                     ! go a specified amount of time before/after
+                     ! climatological date
+                     if (ivt(p)==nsoybean .or. ivt(p) == nsoybeanirrig) gddmaturity(p)=min(gdd1020(p),hybgdd(ivt(p)))
+                     if (ivt(p)==ncorn .or. ivt(p)==ncornirrig) then
+                        gddmaturity(p)=max(950._r8, min(gdd820(p)*0.85_r8, hybgdd(ivt(p))))
```
Comment (bishtgautam) [inline components/clm/src/biogeochem/PhenologyMod.F90]: what do you think about defining new variables for `950._r8`,  `0.85_r8`, `150._r8`, and `1850._r8`?

File: components/clm/src/biogeochem/CNPhenologyBeTRMod.F90 [hunk 4]
```diff
@@ -1542,45 +1576,78 @@ subroutine CropPhenology(num_pcropp, filter_pcropp                     , &
...
+                     idop(p)      = jday
+                     harvday(p)   = NOT_Harvested
+                     harvdate(p)  = NOT_Harvested
+                     harvdate(p)  = NOT_Harvested
+
+                     ! go a specified amount of time before/after
+                     ! climatological date
+                     if (ivt(p)==nsoybean .or. ivt(p) == nsoybeanirrig) gddmaturity(p)=min(gdd1020(p),hybgdd(ivt(p)))
+                     if (ivt(p)==ncorn .or. ivt(p)==ncornirrig) then
+                        gddmaturity(p)=max(950._r8, min(gdd820(p)*0.85_r8, hybgdd(ivt(p))))
```
Comment (bishtgautam) [inline components/clm/src/biogeochem/CNPhenologyBeTRMod.F90]: same a previous comment: what do you think about defining new variables for `950._r8`,  `0.85_r8`, `150._r8`, and `1850._r8`?

File: components/clm/src/biogeochem/CropMod.F90 [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    ! !DESCRIPTION
+    ! calculates the plant month based on the type of seasonality
+
+    ! !USES:
+    use pftvarcon        , only : planttemp
+    ! !ARGUMENTS:
+    implicit none
+    integer,  intent(in)  :: p
+    real(r8), intent(in)  :: cvt      ! CV of temperature (K)
```
[the remark this answers is not in the document]
Comment (bbye) [reply components/clm/src/biogeochem/CropMod.F90]: cvt, cvp and p2e are updated annually. temp is updated each timestep. All of them are an average, but it's an exponential weighted moving average (not a simple average). This subroutine is only called once a year.

File: components/clm/src/biogeochem/CropMod.F90 [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    i = 1
+    t_m = indx(1)
+    do while ((temp(i) .lt. planttemp(ivt(p)) .or. jday(i) .lt. minplantjday) .and. i .lt. 12)
+        i = i+1
+        t_m = indx(i) ! the reason this is set after the index count update is to make
+                      ! sure the plant month is the first month temp > planttemp
+    end do
+    if (cvp > tcvp) then
+       if (cvt >= tcvt) then    ! Both temperature and precip seasonality
+          if (minval(temp) .lt. 283._r8)then ! cold season exists
```
[the remark this answers is not in the document]
Comment (bbye) [reply components/clm/src/biogeochem/CropMod.F90]: OK

File: components/clm/src/biogeochem/CNPhenologyBeTRMod.F90 (hunk 2 above)
Comment (bbye) [reply components/clm/src/biogeochem/CNPhenologyBeTRMod.F90]: OK

File: components/clm/src/biogeochem/PhenologyMod.F90 (hunk 3 above)
Comment (bbye) [reply components/clm/src/biogeochem/PhenologyMod.F90]: I would like to hold off on defining these. They were grandfathered from the CLM crop model, but I think there is some room for modifying these. As part of some crop code cleanup and the parameter calibration process I will revisit these numbers and might pull them into the external parameter file.

File: components/clm/src/biogeochem/CNPhenologyBeTRMod.F90 (hunk 4 above)
Comment (bbye) [reply components/clm/src/biogeochem/CNPhenologyBeTRMod.F90]: See above comment.

File: components/clm/src/biogeochem/CNPhenologyBeTRMod.F90 (hunk 1 above)
Comment (bbye) [reply components/clm/src/biogeochem/CNPhenologyBeTRMod.F90]: OK, I pulled tcvp, tcvt, and cst (the temperature variable 283._r8) into the CropType and defined them as a parameter there.

**Comment by rljacob** [comment]
I see an approved data on June 18 here:  https://acme-climate.atlassian.net/wiki/spaces/ED/pages/994934933/B4+Crop+Plant+Date+Design+Document

@kvcalvin you approved it right?

**Comment by kvcalvin** [comment]
I did approve the design document. Is there something else I need to approve?

**Comment by bishtgautam** [comment]
@rljacob, Do we also need to approval for verification, performance, and validation tasks before we integrate v2 PRs? see https://acme-climate.atlassian.net/wiki/spaces/ED/pages/995426324/B4+Crop+Plant+Date

**Comment by rljacob** [comment]
Reassigning to balance the integrator workload

**Comment by qzhu-lbl** [comment]
@bbye  energyflux_vars was added in subroutine "CNPhenology", in order to be used in "CropPlantDate". However, it seems that energyflux_vars is never used in "CropPlantDate"？ Am I understanding this correctly? Is energyflux_vars redundant?

**Comment by bbye** [comment]
I think I see what happened. The variable I used from the energyflux_vars is defined in both EnergyFluxMod and VegetationDataType. When I switched to use veg_ef I didn't remove the energyflux_vars. I can fix this. But, does anyone know why many of the energyflux_vars are allocated in two subroutines?

**Comment by jqyin** [comment]
@thorntonpe @qzhu-lbl can you approve this PR?

**Comment by jqyin** [comment]
Merged to next.

**Comment by rljacob** [comment]
There were no diffs in the only crop test:  SMS_Ly2_P1x1.1x1_smallvilleIA.ICLM45CNCROP.sandiatoss3_intel.clm-force_netcdf_pio  Is that expected?

**Comment by bbye** [comment]
Yes, that is expected. The changes are minor for the temperature threshold, which is the case for the crop test. In other grid cells there will be changes. In a future PR I will update the parameter file with new values for temperature thresholds that will result in diffs, but I can't do that until #2992 and #2995 are fixed.

**Comment by jqyin** [comment]
Merged #2999

**Pull request merged**